From 5621af0a01cfd03acdac7363d8580af8d6699670 Mon Sep 17 00:00:00 2001
From: Lucas Kanashiro <lucas.kanashiro@canonical.com>
Date: Wed, 23 Aug 2023 18:54:54 -0300
Subject: [PATCH 1/6] snappy-aa-profile-reload

---
 daemon/apparmor_default.go | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/daemon/apparmor_default.go b/daemon/apparmor_default.go
index a1048e303c..55934032ea 100644
--- a/daemon/apparmor_default.go
+++ b/daemon/apparmor_default.go
@@ -4,6 +4,7 @@ package daemon
 
 import (
 	"fmt"
+	"os"
 
 	"github.com/containerd/containerd/v2/pkg/apparmor"
 	aaprofile "github.com/moby/profiles/apparmor"
@@ -31,7 +32,7 @@ func ensureDefaultAppArmorProfile() error {
 		}
 
 		// Nothing to do.
-		if loaded {
+		if loaded && (os.Getenv("DOCKER_AA_RELOAD") == "" || os.Getenv("DOCKER_AA_RELOAD") == "0") {
 			return nil
 		}
 
-- 
2.43.0

