diff --git a/vendor/github.com/moby/buildkit/source/git/gitsource.go b/vendor/github.com/moby/buildkit/source/git/gitsource.go
index dd35fe55f7..466c0720ca 100644
--- a/vendor/github.com/moby/buildkit/source/git/gitsource.go
+++ b/vendor/github.com/moby/buildkit/source/git/gitsource.go
@@ -672,6 +672,12 @@ func git(ctx context.Context, dir, sshAuthSock, knownHosts string, args ...strin
 			"GIT_CONFIG_NOSYSTEM=1", // Disable reading from system gitconfig.
 			"HOME=/dev/null",        // Disable reading from user gitconfig.
 		}
+		if envVal, ok := os.LookupEnv("GIT_EXEC_PATH"); ok {
+			cmd.Env = append(cmd.Env, "GIT_EXEC_PATH="+envVal)
+		}
+		if envVal, ok := os.LookupEnv("GIT_TEMPLATE_DIR"); ok {
+			cmd.Env = append(cmd.Env, "GIT_TEMPLATE_DIR="+envVal)
+		}
 		if sshAuthSock != "" {
 			cmd.Env = append(cmd.Env, "SSH_AUTH_SOCK="+sshAuthSock)
 		}
