diff --git a/builder/remotecontext/git/gitutils.go b/builder/remotecontext/git/gitutils.go
index d6e3d19355..7026fdf223 100644
--- a/builder/remotecontext/git/gitutils.go
+++ b/builder/remotecontext/git/gitutils.go
@@ -215,6 +215,13 @@ func (repo gitRepo) gitWithinDir(dir string, args ...string) ([]byte, error) {
 	// Disable unsafe remote protocols.
 	cmd.Env = append(cmd.Env, "GIT_PROTOCOL_FROM_USER=0")
 
+	if envVal, ok := os.LookupEnv("GIT_EXEC_PATH"); ok {
+		cmd.Env = append(cmd.Env, "GIT_EXEC_PATH="+envVal)
+	}
+	if envVal, ok := os.LookupEnv("GIT_TEMPLATE_DIR"); ok {
+		cmd.Env = append(cmd.Env, "GIT_TEMPLATE_DIR="+envVal)
+	}
+
 	if repo.isolateConfig {
 		cmd.Env = append(cmd.Env,
 			"GIT_CONFIG_NOSYSTEM=1", // Disable reading from system gitconfig.
